File: phpoc-web/src/sync/migration.js
/**
 * Migration: blob-to-rows migration for staging data.
 *
 * Converts the old single-blob `entries` format (array of {hash, data, committed})
 * to the new row-level staging store (`staging:row:{activity_id}` keys).
 *
 * Provides:
 *   migrateBlobToRows(storage) → Promise<number> — runs migration, returns row count
 *
 * Design:
 *   - Idempotent: checks `staging:migration:row_level` marker, skips if present
 *   - O(1) marker check before all operations (fast exit when already done)
 *   - Best-effort: corrupted entries are skipped, good entries are migrated
 *   - Removes old `entries` key after successful completion
 *   - Generates activity_id for entries that lack one via simple random ID
 *   - Sets updated_at from current timestamp since old format has no concept of it
 */

import { RowStagingStore } from './row_staging_store.js';
import { generateActivityId } from './activity_id.js';

const MIGRATION_MARKER_KEY = 'staging:migration:row_level';
const ENTRIES_KEY = 'entries';

/**
 * Run the blob-to-rows migration for staging data.
 *
 * @param {import('./storage.js').StorageBackend} storage — Storage backend.
 * @param {object} [options]
 * @param {() => string} [options.generateId] — Injectible activity_id generator (test seam).
 * @returns {Promise<number>} Number of rows migrated (0 if already done or empty).
 */
export async function migrateBlobToRows(storage, options = {}) {
  // Fast exit: marker already present
  const marker = await storage.get(MIGRATION_MARKER_KEY);
  if (marker !== undefined && marker !== null) {
    return 0;
  }

  const generateId = options.generateId || generateActivityId;

  // Read old blob entries
  const entries = (await storage.get(ENTRIES_KEY)) || [];
  if (!Array.isArray(entries) || entries.length === 0) {
    // Write marker even for empty migration so we don't re-check
    await storage.set(MIGRATION_MARKER_KEY, true);
    return 0;
  }

  const rowStore = new RowStagingStore(storage);
  const now = Date.now();
  let migrated = 0;

  for (const entry of entries) {
    try {
      if (!entry || !entry.data) continue;

      const data = entry.data || {};

      // Use existing activity_id (entry-level or data-level) or generate one
      const activityId = entry.activity_id || data.activity_id || generateId();

      // Determine status from flags in old format
      let activityStatus = 'staged';
      if (entry.committed) {
        activityStatus = 'ended';
      } else if (data.is_active && data.is_paused) {
        activityStatus = 'paused';
      } else if (data.is_active) {
        activityStatus = 'active';
      }

      // Build activity blob preserving all old data
      const activityBlob = {
        entry_id: data.entry_id || '',
        title: data.title || '',
        startTime_enc: data.startTime_enc || '',
        endTime_enc: data.endTime_enc || undefined,
        duration: data.duration || 0,
        is_active: data.is_active,
        is_paused: data.is_paused,
        pauses_enc: data.pauses_enc || 'plain:[]',
        tags: data.tags || [],
        media: data.media || [],
        device_uuid_enc: data.device_uuid_enc || '',
        end_device_uuid_enc: data.end_device_uuid_enc || '',
        metadata_enc: data.metadata_enc || 'plain:{}',
        comment: data.comment || null,
        committed: entry.committed || false,
        block_index: entry.block_index ?? null,
        hash: entry.hash || '',
      };

      // Remove undefined keys
      for (const k of Object.keys(activityBlob)) {
        if (activityBlob[k] === undefined) delete activityBlob[k];
      }

      const row = {
        activity_id: activityId,
        activity_status: activityStatus,
        activity: JSON.stringify(activityBlob),
        updated_at: now,
      };

      await rowStore.putRow(row);
      migrated++;
    } catch {
      // Best-effort: skip corrupted entries
      continue;
    }
  }

  // Remove old entries key after successful migration
  await storage.delete(ENTRIES_KEY);

  // Write marker so subsequent calls are no-ops
  await storage.set(MIGRATION_MARKER_KEY, true);

  return migrated;
}

File: phpoc-flutter/lib/data/storage/migration.dart
import 'dart:convert';

import '../storage/database.dart';
import '../sync/activity_id.dart';
import '../sync/staging_store.dart';

/// Migrates old `entries` JSON-array blob → new row-per-activity `staging` table.
///
/// One-shot migration that:
///   1. Reads the legacy `entries` key from the old K/V store
///   2. Converts each entry to a `staging` row
///   3. Writes all rows to the new StagingStore
///   4. Sets a `migrated_v1` flag to prevent re-run
///   5. Deletes the old `entries` key
class StagingMigration {
  final AppDatabase db;
  final dynamic legacyStorage; // {get, set, remove, hasKey}
  final StagingStore stagingStore;

  StagingMigration({
    required this.db,
    required this.legacyStorage,
    required this.stagingStore,
  });

  /// Check whether migration is needed.
  ///
  /// Returns true when the legacy `entries` key exists AND the staged
  /// table is empty (or nearly empty).
  Future<bool> needsMigration() async {
    // If staging table already has rows, no migration needed (C2)
    final count = await stagingStore.count();
    if (count > 0) return false;

    // Check for legacy entries
    try {
      final entries = await legacyStorage.get('entries');
      if (entries is List && entries.isNotEmpty) return true;
    } catch (_) {
      return false;
    }

    // Check for migration marker
    try {
      final migrated = await legacyStorage.get('migrated_v1');
      if (migrated == true) return false;
    } catch (_) {}

    return false;
  }

  /// Run the migration. Idempotent — safe to call multiple times.
  Future<void> migrate() async {
    // Double-check needs migration (C2)
    if (!await needsMigration()) return;

    // Read legacy entries
    List legacyEntries;
    try {
      final raw = await legacyStorage.get('entries');
      if (raw is! List) return; // C11: malformed blob
      legacyEntries = raw;
    } catch (_) {
      return; // C11
    }

    // C10: empty blob
    if (legacyEntries.isEmpty) {
      await _cleanupLegacy();
      return;
    }

    final now = DateTime.now().millisecondsSinceEpoch;

    for (final raw in legacyEntries) {
      if (raw is! Map) continue;

      // C8: skip committed entries
      if (raw['committed'] == true) continue;

      final data = raw['data'] as Map<String, dynamic>? ?? {};

      // C9: preserve existing activity_id (from web-originated entries)
      final existingId = data['entry_id'] as String?;
      final activityId = (existingId != null && existingId.isNotEmpty)
          ? existingId
          : ActivityIdGenerator.generateActivityId();

      // C4: derive status from is_active/is_paused
      final isActive = data['is_active'] as bool? ?? true;
      final isPaused = data['is_paused'] as bool? ?? false;
      String activityStatus;
      if (!isActive) {
        activityStatus = 'ended';
      } else if (isPaused) {
        activityStatus = 'paused';
      } else {
        activityStatus = 'active';
      }

      // C5: preserve original encrypted data blob as JSON
      final activity = json.encode(data);

      // C6: extract fields from data into row-level extras for commit() compat.
      // Old KV entries use startTime_enc (encrypted), not start_epoch (plaintext).
      // Only extract fields that are already plaintext.
      final startEpoch = data['start_epoch'];
      final entryEndEpoch = data['end_epoch'];

      // C7: use migration time for updated_at
      await stagingStore.putRow({
        'activity_id': activityId,
        'activity_status': activityStatus,
        'activity': activity,
        'updated_at': now,
        if (startEpoch is int && startEpoch > 0) 'start_epoch': startEpoch,
        'title': data['title'] ?? '',
        'duration': data['duration'] ?? 0,
        if (entryEndEpoch is int) 'end_epoch': entryEndEpoch,
        'tags': data['tags'] ?? [],
        'pauses': data['pauses'] ?? [],
        if (data['comment'] != null) 'comment': data['comment'],
      });
    }

    // C8: set migration marker
    try {
      await legacyStorage.set('migrated_v1', true);
    } catch (_) {}

    // C12: delete old entries key
    await _cleanupLegacy();
  }

  Future<void> _cleanupLegacy() async {
    try {
      await legacyStorage.remove('entries');
    } catch (_) {}
  }
}
